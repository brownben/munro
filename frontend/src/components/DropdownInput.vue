<!--
  Dropdown Form Input

  Dropdown input for forms wrapper for the select element.
  'Emits' a value when changed
-->

<template>
  <div class="dropdown-input">
    <div class="visible" @click="toggle">
      <label>{{ label }}</label>
      <p>{{ currentValue }}</p>
      <svg :class="{ active: open }" fill="#9E9E9E" height="24" viewBox="0 0 24 24" width="24">
        <path d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z" />
        <path d="M0-.75h24v24H0z" fill="none" />
      </svg>
    </div>
    <transition name="open">
      <div v-show="open" class="dropdown">
        <p v-for="item in list" :key="item" @click="changeSelection(item)">{{ item }}</p>
      </div>
    </transition>
  </div>
</template>

<script>
export default {
  name: 'DropdownInput',

  props: {
    'label': {
      type: String,
      default: '',
    },
    'value': {
      type: String,
      default: '',
    },
    'list': {
      type: Array,
      default: () => [],
    },
  },

  data: function () {
    return {
      open: false,
      currentValue: this.value,
    }
  },

  watch: {
    value: function (value) {
      this.currentValue = value
    },
  },

  methods: {
    changeSelection: function (value) {
      this.open = false
      this.currentValue = value
      this.$emit('input', value)
    },

    toggle: function () {
      this.open = !this.open
      if (this.open) this.$emit('opened')
    },
  },
}
</script>

<style scoped lang="stylus">
@import '../assets/styles/helpers'

.dropdown-input
  display: block
  box-sizing: border-box
  padding: 0.2rem 0
  width: 100%
  height: 1.84rem
  outline: 0
  border: main-color solid 1px
  background-color: white
  font-weight: 300
  font-size: 1rem
  font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif
  user-select: none

  &:last-child
    border-bottom: 0

.visible
  position: relative

  p
    display: inline-block
    padding-left: 0.25rem
    color: black
    font-size: 1rem

  label
    display: inline-block
    width: auto
    color: main-color
    font-size: 1rem

  svg
    position: absolute
    top: -0.1rem
    right: 0

    .active
      fill: main-color

.dropdown
  position: relative
  top: 0.2rem
  left: -0.05rem
  z-index: 2
  box-sizing: border-box
  width: calc(100% + 0.1rem)
  border: 1px solid alpha(main-color, 0.4)
  background-color: white
  color: black
  font-size: 1rem

  p
    display: block
    padding: 0.25rem

    &:hover
      background-color: #E1BEE7

.open-enter-active, .open-leave-active
  transition: 0.3s transform
  transform: scaleY(1)
  transform-origin: top

.open-enter, .open-leave-to
  transform: scaleY(0)
  transform-origin: top
</style>
