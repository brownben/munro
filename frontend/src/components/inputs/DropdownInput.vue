<!--
  Dropdown Form Input

  Dropdown input for forms wrapper for the select element.
  'Emits' a value when changed
-->

<template>
  <div>
    <label
      class="block pb-1 text-gray-600 select-none font-heading"
      :for="label"
    >
      {{ label }}
    </label>

    <div class="relative">
      <span
        class="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
      >
        <svg
          class="w-5 h-5 text-gray-400"
          viewBox="0 0 24 24"
          fill="none"
          stroke="currentColor"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-width="1.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>

      <select
        :id="label"
        v-model="currentValue"
        class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main"
        @change="$emit('input', currentValue)"
      >
        <option v-if="shift" />
        <template v-if="!optionTextDifferentToValue">
          <option v-for="item in list" :key="item" :value="item">
            {{ item }}
          </option>
        </template>
        <template v-else>
          <option v-for="item in list" :key="item.text" :value="item.value">
            {{ item.text }}
          </option>
        </template>
      </select>
    </div>
  </div>
</template>

<script>
export default {
  name: 'DropdownInput',

  props: {
    label: {
      type: String,
      default: '',
    },
    value: {
      type: String,
      default: '',
    },
    list: {
      type: Array,
      default: () => [],
    },
    shift: {
      type: Boolean,
      default: true,
    },
    optionTextDifferentToValue: {
      type: Boolean,
      default: false,
    },
  },

  data: function () {
    return {
      open: false,
      currentValue: this.value,
    }
  },

  watch: {
    value: function (value) {
      this.currentValue = value
    },
  },
}
</script>

<style lang="postcss">
option {
  @apply bg-white;
}
</style>
