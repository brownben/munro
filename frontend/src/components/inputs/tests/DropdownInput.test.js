import { shallowMount } from '@vue/test-utils'
import DropdownInput from '../DropdownInput'

test('Renders Correctly', () => {
  const wrapper = shallowMount(DropdownInput, {
    props: {
      label: 'Test Label:',
      type: 'text',
    },
  })
  expect(wrapper.exists()).toBeTruthy()
  expect(wrapper.find('label').text()).toBe('Test Label:')
  expect(wrapper.find('select').exists()).toBeTruthy()
  expect(wrapper.html()).toContain('type="text"')
})

test('Focus Is Correctly Applied to Label', async () => {
  const wrapper = shallowMount(DropdownInput)
  expect(wrapper.find('label').classes('text-main-600')).toBeFalsy()
  expect(wrapper.find('label').classes('text-gray-600')).toBeTruthy()
  await wrapper.find('select').trigger('focus')
  expect(wrapper.find('label').classes('text-main-600')).toBeTruthy()
  expect(wrapper.find('label').classes('text-gray-600')).toBeFalsy()
  await wrapper.find('select').trigger('blur')
  expect(wrapper.find('label').classes('text-main-600')).toBeFalsy()
  expect(wrapper.find('label').classes('text-gray-600')).toBeTruthy()
})

test('Emits Data On Change', async () => {
  const wrapper = shallowMount(DropdownInput, {
    props: {
      list: ['1', 'newOutput', '3'],
      shift: false,
    },
  })
  await wrapper.find('select').setValue('newOutput')
  expect(wrapper.emitted()).toHaveProperty('update:modelValue')
  expect(wrapper.emitted()['update:modelValue'][0]).toContain('newOutput')
})

test('Shift Option Works', () => {
  const wrapper = shallowMount(DropdownInput, {
    props: { shift: true },
  })
  const options = wrapper.findAll('option')
  expect(options.length).toBe(1)
  expect(options[0].text()).toBe('')
})

test('String Options', () => {
  const wrapper = shallowMount(DropdownInput, {
    props: {
      list: ['1', '2', '3'],
      shift: false,
    },
  })
  const options = wrapper.findAll('option')
  expect(options.length).toBe(3)
  expect(options.map((option) => option.text())).toEqual(['1', '2', '3'])
})

test('Object Options', () => {
  const wrapper = shallowMount(DropdownInput, {
    props: {
      list: ['1', '2', '3'].map((item) => ({
        text: item,
        value: parseInt(item, 10),
      })),
      shift: false,
      optionTextDifferentToValue: true,
    },
  })
  const options = wrapper.findAll('option')
  expect(options.length).toBe(3)
  expect(options.map((option) => option.text())).toEqual(['1', '2', '3'])
  expect(options.map((option) => option.element._value)).toEqual([1, 2, 3])
})
