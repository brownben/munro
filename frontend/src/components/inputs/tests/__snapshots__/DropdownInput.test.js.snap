// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<div
  includeblank="true"
>
  <label
    class="block pb-1 text-gray-600 select-none font-heading"
    for=""
  >
    
    
  
  </label>
   
  <div
    class="relative"
  >
    <span
      class="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
    >
      <svg
        class="w-5 h-5 text-gray-400"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </span>
     
    <select
      class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main"
      id=""
    >
      <option />
       
      <option
        value="Item 1"
      >
        
          Item 1
        
      </option>
      <option
        value="Item 2"
      >
        
          Item 2
        
      </option>
      <option
        value="Item 3"
      >
        
          Item 3
        
      </option>
    </select>
  </div>
</div>
`;
